Provide access to the features as per their feature stage

lkbhitesh07 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of N/A.
2. This PR does the following: This PR handles the configuration of the features as per their feature-stage.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
**Here for the testing purposes I have passed the server stage as the "prod"**

https://github.com/oppia/oppia/assets/44300735/dabc0acb-4393-43ff-84ab-441ec92f09e3


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

lkbhitesh07 committed:
initial changes

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
fixed tests

lkbhitesh07 committed:
fixed e2e test

lkbhitesh07 committed:
fixed lint

lkbhitesh07 committed:
fixed the failing tests

lkbhitesh07 committed:
ts check fix

lkbhitesh07 committed:
fixed frontend tests

lkbhitesh07 commented:
Hey @seanlip, @gp201, @chris7716 PTAL.

seanlip commented on core/tests/webdriverio/featureGatingFlow.js:
```diff
@@ -54,80 +54,71 @@ describe('Feature Gating Flow', function() {
-      expect(await ajsDummyFeatureIndicator.isExisting()).toBe(false);
-      await users.logout();
-    });
+  it('should not show indicators for contributor_dashboard_accomplishments ' +
```
Ah, ignore my suggestion elsewhere to deprecate the dummy platform parameter. I guess it's used for e2e tests.

But why don't you just make that dummy platform parameter a feature flag so you can use it in this way? Otherwise there'll be churn here when we get rid of the CD flag once the feature is confirmed to work well.

lkbhitesh07 replied:
Oh, well we already have a dummy feature flag and we are using it in this file only, i.e. featureGatingFlow. The reason why I have to replace the dummy feature to the CD flag is that the e2e tests run in prod mode and the dummy feature flag has the feature stage as "dev" hence the tests were failing as the dummy feature is not accessible in the prod. So, I think we would require to make the changes when we get rid of the CD flag.
Thanks

seanlip replied:
Oh, I see. So what's the long-term plan here then? The CD flag is likely to be removed soon.

Does the dummy flag result in any user-visible change? If not I would rename it to dummy_feature_flag_for_e2e_tests and just use that instead (if all it does is introduce an "e2e" class that has no other effect).

lkbhitesh07 replied:
If you agree, can we change the feature stage of dummy-feature from "dev" to "prod"? The reason for this would be that at some point we will not have any feature flags as all of them will be implemented permanently in the codebase and then this e2e test will going to fail. Let me know what you think about this or if you have other alternatives please let me know.

Well, we do have a div for the dummy feature, have a look at the bottom-left in the screenshot:
![Screenshot from 2023-06-17 22-38-22](https://github.com/oppia/oppia/assets/44300735/9a8eb194-9180-4fef-8c0c-eeb6e4c336e5)

seanlip reviewed (changes requested):
Thanks @lkbhitesh07, looks pretty good. In terms of UI can we just get rid of the rules section for disabled features, and show them at the botttom of the list?

I think you can also drop the "feature stage" indicator for those and have the error message at the top say something like "This feature flag is in the XXX stage, and cannot be configured on this server (which is in the 'YYY' server mode)".

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
updates to the features tab

lkbhitesh07 committed:
bit

lkbhitesh07 commented:
> Thanks @lkbhitesh07, looks pretty good. In terms of UI can we just get rid of the rules section for disabled features, and show them at the botttom of the list?
> 
> I think you can also drop the "feature stage" indicator for those and have the error message at the top say something like "This feature flag is in the XXX stage, and cannot be configured on this server (which is in the 'YYY' server mode)".

Thanks for the review @seanlip, I have made the changes and updated the PR description.
@seanlip PTAL

lkbhitesh07 commented:
Thanks @seanlip, [Replied](https://github.com/oppia/oppia/pull/18386#discussion_r1233097499) to your comment.
Thanks

seanlip commented:
You can change it to prod, but only if it is modified to have no user
visible effects.

On Sun, Jun 18, 2023, 1:20 AM Hitesh Tomar ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In core/tests/webdriverio/featureGatingFlow.js
> <https://github.com/oppia/oppia/pull/18386#discussion_r1233097499>:
>
> > -      // is enabled, and the backend dummy handler is also enabled.
> -      var agDummyHandlerIndicator = $(
> -        '.e2e-test-angular-dummy-handler-indicator');
> -
> -      // Indicator in AngularJS directive that is visible if the dummy_feature
> -      // is enabled, and the feature status is successfully loaded in the
> -      // AngularJS directive.
> -      var ajsDummyFeatureIndicator = $(
> -        '.e2e-test-angularjs-dummy-feature-indicator');
> -
> -      expect(await agDummyFeatureIndicator.isExisting()).toBe(false);
> -      expect(await agDummyHandlerIndicator.isExisting()).toBe(false);
> -      expect(await ajsDummyFeatureIndicator.isExisting()).toBe(false);
> -      await users.logout();
> -    });
> +  it('should not show indicators for contributor_dashboard_accomplishments ' +
>
> If you agree, can we change the feature stage of dummy-feature from "dev"
> to "prod"? The reason for this would be that at some point we will not have
> any feature flags as all of them will be implemented permanently in the
> codebase and then this e2e test will going to fail. Let me know what you
> think about this or if you have other alternatives please let me know.
>
> Well, we do have a div for the dummy feature, have a look at the
> bottom-left in the screenshot:
> [image: Screenshot from 2023-06-17 22-38-22]
> <https://user-images.githubusercontent.com/44300735/246622844-9a8eb194-9180-4fef-8c0c-eeb6e4c336e5.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/18386#discussion_r1233097499>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQV5SS4XHYVWPGNHIAXAGTXLXRORANCNFSM6AAAAAAZHHTQMM>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

lkbhitesh07 committed:
changed feature stage of dummy-feature

lkbhitesh07 commented:
Thanks @seanlip, I have made the necessary changes.
Thanks

lkbhitesh07 committed:
fixed backend test

seanlip commented:
@lkbhitesh07 I don't see the changes I requested [here](https://github.com/oppia/oppia/pull/18386#discussion_r1233077635) (re renaming of the flag), and I'm not sure how to audit the user-visible part. What did you change about the latter?

(If the flag's effects are user-visible only on the /admin page then I guess that's fine too.)

gp201 commented on core/templates/pages/release-coordinator-page/features-tab/features-tab.component.spec.ts:
```diff
@@ -313,6 +314,91 @@ describe('Release coordinator page feature tab', function() {
+
+    it('should return false when the server in test stage and feature ' +
+    'stage is dev', (() => {
+      component.serverStage = 'test';
+      expect(component.getFeatureValidOnCurrentServer(
+        featureFlagDevStage)).toBe(false);
```
```suggestion
      component.serverStage = 'test';
 
      expect(component.getFeatureValidOnCurrentServer(
        featureFlagDevStage)).toBe(false);
```

Please make the same change for the remaining tests. Thanks!

lkbhitesh07 replied:
Done

gp201 reviewed (changes requested):
Thanks, @lkbhitesh07 ! I left some comments. PTAL

lkbhitesh07 commented:
Hey @seanlip,
Re your comment [here](https://github.com/oppia/oppia/pull/18386#discussion_r1233077635): For the first part which is the fix for the long-term plan, I have changed the dev stage of the dummy-feature from "dev" to "prod" and reverted the changes I did in the featureGating file. We will be using the dummy-feature for the e2e test as we did previously.

For the second part: The dummy-feature has a user-facing effect and that is only visible on the admin page(I have posted the screenshot on the comment thread) and that is why I did not change the name to dummy_feature_flag_for_e2e_tests as we do have a user-facing effect.
Let me know if the name still needs to be changed.
Thanks

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
added blank line to the spec file

lkbhitesh07 commented:
Done with the changes @gp201.
Thanks

seanlip commented:
@lkbhitesh07 Yes, please change the name, since the only reason we really have that flag is for the e2e tests. Thanks.

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

chris7716 commented on core/templates/pages/contributor-dashboard-page/contributions-and-review/contributions-and-review.component.html:
```diff
@@ -30,7 +30,7 @@
           </div>
         </div>
-        <div class="oppia-contributions-option-group-label"
+        <div class="oppia-contributions-option-group-label e2e-test-accomplishments-tab-enabled-indicator"
```
@lkbhitesh07 could you please explain why this change was required?

lkbhitesh07 replied:
Ah! Forgot to remove this part. I actually edited one of the e2e test file and this was required there but I reverted those changes so this change is not needed anymore.
I have removed it, thanks for catching.

chris7716 reviewed (commented):
Added a comment. Thanks!

lkbhitesh07 committed:
renamed dummy feature flag

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
fixed lint

lkbhitesh07 committed:
nit

lkbhitesh07 committed:
nit

lkbhitesh07 commented:
> @lkbhitesh07 Yes, please change the name, since the only reason we really have that flag is for the e2e tests. Thanks.

Done @seanlip.
Thanks

lkbhitesh07 committed:
removed extra dummy feature present on the admin page

lkbhitesh07 committed:
nit fix

seanlip commented on core/templates/pages/release-coordinator-page/features-tab/features-tab.component.ts:
```diff
@@ -316,12 +336,12 @@ export class FeaturesTabComponent implements OnInit {
 
   async reloadDummyHandlerStatusAsync(): Promise<void> {
-    if (this.isDummyFeatureEnabled) {
+    if (this.isDummyFeatureFlagForE2ETestsEnabled) {
```
Is this a function? If so then shouldn't it have a () after it?

If it's a variable then it should be dummyFeatureFlagForE2eTestsIsEnabled

lkbhitesh07 replied:
Yes, it is a function but the function is of type `get` so we would not require to have a (), this function is defined just above this line.
Thanks

seanlip replied:
OK, then are we treating it as a property of `this`? If so my suggested renaming would apply.

lkbhitesh07 replied:
Yes, make sense to rename it. I have updated this.
Thanks

seanlip reviewed (changes requested):
One minor nit, but otherwise no concerns. Thanks!

lkbhitesh07 committed:
Merge branch 'develop' of github.com:oppia/oppia into manage-features-acc-to-feature-stage

lkbhitesh07 committed:
renamed variable

seanlip reviewed (approved):
Thanks, LGTM!

gp201 reviewed (approved):
LGTM for codeowner files

lkbhitesh07 committed:
fixed broken tests

chris7716 reviewed (approved):
LGTM. Thanks!
